File: ownlang/renders.py
"""Layer 3, rendered: the bridge's output surfaces (P-022 #259 checkpoint 5.3).

`ownlang/verdicts.py` freezes what the bridge CONCLUDES — the `Finding` list.
This module freezes what a consumer SEES: the same list through
`ownir.render_finding` (the human CLI line, the GitHub Actions annotation, the
MSBuild/VS Error List line) and through `ownir.build_sarif` (the one SARIF
2.1.0 log per run). Those are BR-V9's surfaces, and until this family they had
no golden of any kind — `tests/test_ownir.py` pinned a handful of their rules
against hand-written strings, which is not the same thing as freezing the bytes
a run emits.

Strictly an OBSERVER, like `ownlang/verdicts.py` and `ownlang/lowered.py`: it
never mutates facts, never changes a verdict or a rendering, and is imported by
nothing in the production path. It calls the production renderers and records
what they returned.

Normalization decisions (frozen; changing any is a parity-contract change):

* **Both host severities, always.** `severity` is a presentation choice that
  the CLI passes through (`--severity`), and it is the one input to these
  surfaces besides the findings. Every case renders at `"error"` and at
  `"warning"` so the pass-through and the SARIF level mapping (BR-V6: an
  advisory stays `note`, an intrinsic warning stays `warning`, a provable leak
  drops from `error` to `warning`) are both visible in the golden.
* **One entry per finding, in the bridge's order** (BR-V8), for the three
  line-per-finding surfaces. The list is never re-sorted here.
* **The unknown format is rendered, not assumed.** `render_finding` falls back
  to the human line for any format it does not know; the golden carries that
  rendering under its own key rather than a claim that it equals `human`.
* **SARIF is carried as the value `build_sarif` returned**, re-serialized by
  this module's own `json.dumps`. The log is a JSON document, not a byte
  stream the reference emits, so key ORDER is what the golden pins (Python
  dicts preserve insertion order, and the Rust replay reproduces it).
* **A refusal** (`OwnIRError` from `check_facts`) projects as
  `{"renders_version": ..., "error": "<message>"}` — the same shape the Layer 3
  verdict surface uses, because a document the bridge refuses has no findings
  to render and the refusal is what a consumer would see.
* Rendering is `json.dumps(indent=2, ensure_ascii=False)` + a trailing newline;
  regeneration is deterministic for identical input.

What this surface deliberately does NOT carry: a diagnostic's `subject`.
`ownir.Finding` has no such member, so no bridge surface can serialize one —
and `tests/test_verdict_render_fixtures.py` asserts that over the rendered
bytes rather than restating it (the checkpoint-4 note left that re-check to
this checkpoint).

Fixture sharing: `tests/fixtures/verdict_renders/manifest.json` is the frozen
case ledger. Each case names its own facts document beside the manifest and the
BR-V9 rules it pins, which is what `tests/verdict_surface_inventory.py` reads
to say whether a rule has a control.
"""

from __future__ import annotations

import json
from typing import Any

from .ownir import OwnIRError, build_sarif, check_facts, render_finding

# The rendered-surface version. Bump on ANY normalization change above — the
# committed goldens and the Rust replay are both keyed to it.
RENDERS_VERSION = 1

# The host severities every case is rendered at (see the docstring).
SEVERITIES: tuple[str, ...] = ("error", "warning")
# The `render_finding` formats, plus one the function does not know: the
# fallback is a rule, so it is rendered rather than asserted.
FORMATS: tuple[str, ...] = ("human", "github", "msbuild", "unknown-format")


def project_renders(facts: dict[str, Any]) -> dict[str, Any]:
    """Project one facts document's rendered surfaces into the canonical dict.
    A bridge refusal projects as `{"renders_version": ..., "error": <message>}`.
    Never mutates `facts`."""
    try:
        findings = check_facts(facts)
    except OwnIRError as e:
        return {"renders_version": RENDERS_VERSION, "error": str(e)}
    out: dict[str, Any] = {"renders_version": RENDERS_VERSION}
    for fmt in FORMATS:
        out[fmt] = {
            severity: [render_finding(f, fmt, severity) for f in findings]
            for severity in SEVERITIES
        }
    out["sarif"] = {severity: build_sarif(findings, severity) for severity in SEVERITIES}
    return out


def render_renders(facts: dict[str, Any]) -> str:
    """The canonical serialized form: fixed key order, 2-space indent, non-ASCII
    preserved, trailing newline. Byte-identical on re-run."""
    return json.dumps(project_renders(facts), indent=2, ensure_ascii=False) + "\n"

File: rust/crates/own-bridge/tests/renders.rs
//! The checkpoint-5.3 acceptance contract (#259): for every case of the
//! rendered-surface fixture family,
//!
//! ```text
//! facts.json → OwnIr → own_bridge::check_facts → render_finding / build_sarif
//!            ≡  <case>.renders.json   BYTE FOR BYTE
//! ```
//!
//! The golden is what the reference's own renderers returned
//! (`ownlang/renders.py`, regenerate: `python
//! tests/test_verdict_render_fixtures.py --write`). This replay reconstructs
//! the whole document — every format at both host severities, the SARIF log at
//! both — serializes it with the same conventions the emitter fixes (2-space
//! indent, non-ASCII preserved, a trailing newline) and compares the BYTES.
//! Not a value comparison: SARIF key order is part of the surface here, and a
//! value comparison would score a reordered log as agreement.
//!
//! Independently enforced, not outsourced to Python:
//! * the ledger is the tree — every manifest case has facts and a golden, and
//!   every facts/golden file on disk is a manifest case;
//! * every case pins at least one BR-V9 ledger row, and the rows the manifest
//!   claims are the rows the inventory reads;
//! * no rendered surface carries a diagnostic `subject` (the checkpoint-4
//!   subject tail, re-checked over the bytes on this side too);
//! * every case is deterministic.

#![allow(clippy::panic, clippy::expect_used, clippy::unwrap_used)]

use own_bridge::{build_sarif, check_facts, render_finding, Finding, SarifLog};
use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;

const FIXDIR: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../../tests/fixtures/verdict_renders"
);

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Manifest {
    #[allow(dead_code)]
    comment: String,
    renders_version: u32,
    cases: Vec<Case>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Case {
    name: String,
    rules: Vec<String>,
    pins: Vec<String>,
}

/// One surface at both host severities. A struct rather than a map because the
/// emitter's KEY ORDER is part of the golden and `serde_json`'s map type sorts.
#[derive(Serialize)]
struct PerSeverity<T> {
    error: T,
    warning: T,
}

/// The whole rendered document, in the emitter's key order
/// (`ownlang/renders.py`). `skip_serializing_if` reproduces the two shapes it
/// emits: a refusal carries `error` and nothing else; a success carries the
/// four rendered formats and the SARIF pair.
#[derive(Serialize)]
struct RenderedDocument {
    renders_version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    human: Option<PerSeverity<Vec<String>>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    github: Option<PerSeverity<Vec<String>>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    msbuild: Option<PerSeverity<Vec<String>>>,
    #[serde(rename = "unknown-format", skip_serializing_if = "Option::is_none")]
    unknown_format: Option<PerSeverity<Vec<String>>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    sarif: Option<PerSeverity<SarifLog>>,
}

fn read(path: &str) -> String {
    std::fs::read_to_string(path).unwrap_or_else(|e| {
        panic!(
            "cannot read {path}: {e} — regenerate: \
             python tests/test_verdict_render_fixtures.py --write"
        )
    })
}

fn stems(suffix: &str) -> BTreeSet<String> {
    let mut out = BTreeSet::new();
    for entry in std::fs::read_dir(FIXDIR).expect("fixture directory is readable") {
        let file = entry.expect("directory entry").file_name();
        let file = file.to_str().expect("fixture filenames are UTF-8");
        if let Some(stem) = file.strip_suffix(suffix) {
            out.insert(stem.to_owned());
        }
    }
    out
}

/// The emitter's serialization: `json.dumps(indent=2, ensure_ascii=False)` plus
/// a trailing newline. `serde_json`'s pretty printer uses the same two-space
/// indent and leaves non-ASCII unescaped, so the two agree on the bytes.
fn serialize(doc: &RenderedDocument) -> String {
    let mut out = serde_json::to_string_pretty(doc).expect("the projection serializes");
    out.push('\n');
    out
}

fn lines_of(findings: &[Finding], fmt: &str) -> PerSeverity<Vec<String>> {
    let render = |severity: &str| {
        findings
            .iter()
            .map(|f| render_finding(f, fmt, severity))
            .collect()
    };
    PerSeverity {
        error: render("error"),
        warning: render("warning"),
    }
}

/// Rebuild one case's whole rendered document.
fn project(facts_text: &str) -> RenderedDocument {
    let empty = RenderedDocument {
        renders_version: 1,
        error: None,
        human: None,
        github: None,
        msbuild: None,
        unknown_format: None,
        sarif: None,
    };
    let facts: own_ir::OwnIr = serde_json::from_str(facts_text)
        .unwrap_or_else(|door| panic!("the typed door refused a rendered-surface case: {door}"));
    match check_facts(&facts) {
        Err(refusal) => RenderedDocument {
            error: Some(refusal.to_string()),
            ..empty
        },
        Ok(findings) => RenderedDocument {
            human: Some(lines_of(&findings, "human")),
            github: Some(lines_of(&findings, "github")),
            msbuild: Some(lines_of(&findings, "msbuild")),
            unknown_format: Some(lines_of(&findings, "unknown-format")),
            sarif: Some(PerSeverity {
                error: build_sarif(&findings, "error"),
                warning: build_sarif(&findings, "warning"),
            }),
            ..empty
        },
    }
}

#[test]
fn replays_every_rendered_surface_byte_for_byte() {
    let manifest: Manifest = serde_json::from_str(&read(&format!("{FIXDIR}/manifest.json")))
        .expect("manifest.json parses (typed, strict)");
    assert_eq!(
        manifest.renders_version, 1,
        "manifest renders_version must match this replay's surface version"
    );

    let mut planned: BTreeSet<String> = BTreeSet::new();
    let mut pinned: BTreeSet<String> = BTreeSet::new();
    for case in &manifest.cases {
        assert!(
            !case.rules.is_empty(),
            "case '{}' must name the BR rules it pins",
            case.name
        );
        assert!(
            !case.pins.is_empty(),
            "case '{}' pins no BR-V9 ledger row — a rendered golden nobody can read a \
             claim off is not evidence",
            case.name
        );
        assert!(
            planned.insert(case.name.clone()),
            "duplicate manifest case name: {}",
            case.name
        );
        pinned.extend(case.pins.iter().cloned());
    }
    assert_eq!(
        planned,
        stems(".facts.json"),
        "manifest case names != *.facts.json under fixtures/verdict_renders"
    );
    assert_eq!(
        planned,
        stems(".renders.json"),
        "planned cases != *.renders.json goldens on disk (missing or orphaned golden)"
    );

    let mut divergences: Vec<String> = Vec::new();
    for name in &planned {
        let facts_text = read(&format!("{FIXDIR}/{name}.facts.json"));
        let ours = serialize(&project(&facts_text));
        assert_eq!(
            ours,
            serialize(&project(&facts_text)),
            "{name}: the rendered projection is not deterministic"
        );
        for marker in ["\"subject\"", "'subject'"] {
            assert!(
                !ours.contains(marker),
                "{name}: a rendered surface carries {marker} — the bridge's Finding has no \
                 diagnostic subject and no output may invent one"
            );
        }
        let golden = read(&format!("{FIXDIR}/{name}.renders.json"));
        if ours != golden {
            let first = golden
                .lines()
                .zip(ours.lines())
                .enumerate()
                .find(|(_, (a, b))| a != b);
            let where_ = first.map_or_else(
                || "the end — one document is longer".to_owned(),
                |(i, (python, rust))| {
                    format!(
                        "line {}\n    python = {python:?}\n    rust   = {rust:?}",
                        i + 1
                    )
                },
            );
            divergences.push(format!("{name}: rendered bytes differ, first at {where_}"));
        }
    }
    assert!(
        divergences.is_empty(),
        "{} rendered-surface divergence(s):\n{}",
        divergences.len(),
        divergences.join("\n")
    );
    eprintln!(
        "cp5.3 rendered surfaces: {} cases replayed byte-for-byte, {} BR-V9 rows pinned",
        planned.len(),
        pinned.len()
    );
}
